src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java

```diff
@@ -658,13 +663,15 @@ final int maxConcurrentServerInitiatedStreams() {
     }
 
     void close() {
-        Log.logTrace("Closing HTTP/2 connection: to {0}", connection.address());
-        if (connection.channel().isOpen()) {
-            GoAwayFrame f = new GoAwayFrame(0,
-                    ErrorFrame.NO_ERROR,
-                    "Requested by user".getBytes(UTF_8));
-            // TODO: set last stream. For now zero ok.
-            sendFrame(f);
+        if (markHalfClosedLocal()) {
+            if (connection.channel().isOpen()) {
+                Log.logTrace("Closing HTTP/2 connection: to {0}", connection.address());
+                GoAwayFrame f = new GoAwayFrame(0,
+                        ErrorFrame.NO_ERROR,
+                        "Requested by user".getBytes(UTF_8));
+                // TODO: set last stream. For now zero ok.
+                sendFrame(f);
+            }
```

Comment:
Unrelated to this PR, even before these changes, looking at this implementation of `close()` it's odd that this method isn't setting `closed` flag to `true` and is neither calling `shutdown` to do the additional things that the `shutdown` does. I haven't yet paid a closer attention to the callers of this method to be sure if that's OK. I'm guessing it is OK, otherwise I think we would have seen issues by now.
Maybe this `close()` method should be renamed to `sendGoAway` if that's what it is meant for?

Reply:
close() is supposed to be a graceful shutdown - sending GOAWAY with no error, and that's why it doesn't close the connection immediately. You are right however that the naming sounds a bit strange. I should probably log another issue to revisit how GOAWAY is handled by the HttpClient in the HTTP/2 stack.

Reply:
> I should probably log another issue to revisit how GOAWAY is handled by the HttpClient in the HTTP/2 stack.

That sounds fine to me.